test/integration/cdi/example-driver/app/kubeletplugin.go

```diff
@@ -0,0 +1,152 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package app
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"os"
+	"path/filepath"
+
+	cdiapi "github.com/container-orchestrated-devices/container-device-interface/pkg/cdi"
+	specs "github.com/container-orchestrated-devices/container-device-interface/specs-go"
+	"k8s.io/klog/v2"
+	drapbv1 "k8s.io/kubernetes/pkg/kubelet/apis/dra/v1alpha1"
+	"k8s.io/kubernetes/test/integration/cdi/example-driver/kubeletplugin"
+)
+
+type examplePlugin struct {
+	logger klog.Logger
+	d      kubeletplugin.DRAPlugin
+
+	cdiDir     string
+	driverName string
+}
+
+var _ drapbv1.NodeServer = &examplePlugin{}
+
+// getJSONFilePath returns the absolute path where CDI file is/should be.
+func (ex *examplePlugin) getJSONFilePath(claimUID string) string {
+	return filepath.Join(ex.cdiDir, fmt.Sprintf("%s-%s.json", ex.driverName, claimUID))
+}
+
+// startPlugin sets up the servers that are necessary for a DRA kubelet plugin.
+func startPlugin(logger klog.Logger, cdiDir, driverName, endpoint, draAddress, pluginRegistrationPath string) (*examplePlugin, error) {
+	ex := &examplePlugin{
+		logger:     logger,
+		cdiDir:     cdiDir,
+		driverName: driverName,
+	}
+
+	// Ensure that directories exist, creating them if necessary. We want
+	// to know early if there is a setup problem that would prevent
+	// creating those directories.
+	if err := os.MkdirAll(ex.cdiDir, os.FileMode(0750)); err != nil {
+		return nil, fmt.Errorf("create CDI directory: %v", err)
+	}
+	if err := os.MkdirAll(filepath.Dir(endpoint), 0750); err != nil {
+		return nil, fmt.Errorf("create socket directory: %v", err)
+	}
+
+	d, err := kubeletplugin.Start(ex.logger, driverName, endpoint, draAddress, pluginRegistrationPath, ex)
+	if err != nil {
+		return nil, fmt.Errorf("start kubelet plugin: %v", err)
+	}
+	ex.d = d
+
+	return ex, nil
+}
+
+// stop ensures that all servers are stopped and resources freed.
+func (ex *examplePlugin) stop() {
+	ex.d.Stop()
+}
+
+// NodePrepareResource ensures that the CDI file for the claim exists. It uses
+// a deterministic name to simplify NodeUnprepareResource (no need to remember
+// or discover the name) and idempotency (when called again, the file simply
+// gets written again).
+func (ex *examplePlugin) NodePrepareResource(ctx context.Context, req *drapbv1.NodePrepareResourceRequest) (*drapbv1.NodePrepareResourceResponse, error) {
+	logger := klog.FromContext(ctx)
+	deviceName := "claim-" + req.ClaimUid
+	kind := ex.driverName + "/test"
+	filePath := ex.getJSONFilePath(req.ClaimUid)
```

Comment:
nit: I'd propose to move this variables to the place where they're used(filePath), or get rid of them (kind, deviceName) if they're used only once.

Reply:
Done. I kept logger where it is (top of the function) because if it was created directly before the first usage and then later another log line got added earlier in the function, the logger initialization would have to be moved.

Reply:
deviceName is used twice.